peer/addrinfo_serde.go

```diff
@@ -19,19 +20,25 @@ func (pi AddrInfo) MarshalJSON() ([]byte, error) {
 
 func (pi *AddrInfo) UnmarshalJSON(b []byte) error {
 	var data map[string]interface{}
-	err := json.Unmarshal(b, &data)
-	if err != nil {
+	if err := json.Unmarshal(b, &data); err != nil {
 		return err
 	}
-	pid, err := IDB58Decode(data["ID"].(string))
-	if err != nil {
-		return err
+	id, ok := data["ID"]
+	if !ok {
+		return errors.New("no peer ID")
+	}
+	if idString, ok := id.(string); ok {
+		pid, err := IDB58Decode(idString)
+		if err != nil {
+			return err
+		}
+		pi.ID = pid
 	}
-	pi.ID = pid
-	addrs, ok := data["Addrs"].([]interface{})
-	if ok {
-		for _, a := range addrs {
-			pi.Addrs = append(pi.Addrs, ma.StringCast(a.(string)))
+	if addrsEntry, ok := data["Addrs"]; ok {
+		if addrs, ok := addrsEntry.([]interface{}); ok {
+			for _, a := range addrs {
+				pi.Addrs = append(pi.Addrs, ma.StringCast(a.(string)))
```

Comment:
PR looks good, the only outstanding issue I see is that `ma.StringCast` will panic if the multiaddr is malformed or we don't know one of the codecs in the multiaddr. Given this might contain an unknown codec perhaps we should just ignore the particular addr and continue trying to unmarshal the others?